src/stash/src/lib.rs

```diff
@@ -9,59 +9,66 @@
 
 //! Durable metadata storage.
 
-use std::cmp;
 use std::error::Error;
 use std::fmt;
 use std::iter;
 use std::marker::PhantomData;
-use std::path::Path;
-use std::sync::{Arc, Mutex};
 
-use rusqlite::{named_params, params, Connection, Transaction};
+use timely::progress::frontier::AntichainRef;
 use timely::progress::Antichain;
-use timely::PartialOrder;
 
 use mz_persist_types::Codec;
-use timely::progress::frontier::AntichainRef;
 
-const APPLICATION_ID: i32 = 0x0872_e898; // chosen randomly
+mod sqlite;
+
+pub use crate::sqlite::Sqlite;
+
+pub type Diff = i64;
+pub type Timestamp = i64;
+pub type Id = i64;
+
+pub trait StashConn {
+    fn collection<K, V>(&self, name: &str) -> Result<StashCollection<Self, K, V>, StashError>
+    where
+        Self: Sized,
+        K: Codec + Ord,
+        V: Codec + Ord;
+
+    fn iter(
+        &self,
+        collection_id: Id,
+    ) -> Result<Vec<((Vec<u8>, Vec<u8>), Timestamp, Diff)>, StashError>;
+
+    fn iter_key(
+        &self,
+        collection_id: Id,
+        key: &[u8],
+    ) -> Result<Vec<(Vec<u8>, Timestamp, Diff)>, StashError>;
 
-const SCHEMA: &str = "
-CREATE TABLE collections (
-    collection_id integer PRIMARY KEY,
-    name text NOT NULL UNIQUE
-);
+    fn update_many<I>(&self, collection_id: Id, entries: I) -> Result<(), StashError>
+    where
+        I: IntoIterator<Item = ((Vec<u8>, Vec<u8>), Timestamp, Diff)>;
+
+    fn seal(&self, collection_id: Id, new_upper: AntichainRef<Timestamp>)
```

Comment:
I have a followup PR that does this by forcing the trait impl to do the hard work: https://github.com/mjibson/materialize/commit/2af43f07d0e538764da0c19fe70650785f713e5d#diff-7de5a5c4a69f77e638b594788b2a3469e7e6c7bbd2cb1735f1027f228ffaca69R236

Reply:
amazing 🙌 ! looks like `update_many` and `seal` are still around in that commit. we won't be able to implement them on the new persist api (which only has append style writes), so I think part of my question was if anyone had a sense for how hard it will be to migrate uses of them to `append`

Reply:
Yeah we're in this exciting world where we have both the old and new api at once and it's not great. I'm going to work on this soon, but I think I can move the existing stash user (just timestamp bindings I think) to this new append api and then remove the old update/seal api which should unblock using storage as a stash impl.

Reply:
coord tables have the same problem but are maybe getting fixed in https://github.com/MaterializeInc/materialize/pull/11572

Reply:
awesome, i'm glad you're on it!